pkg/api/validate/openshiftcluster_validatedynamic.go

```diff
@@ -287,3 +294,29 @@ func (dv *openShiftClusterDynamicValidator) validateVnet(ctx context.Context, vn
 
 	return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidLinkedVNet, "", "The provided vnet '%s' is invalid: custom DNS servers are not supported.", vnetID)
 }
+
+func (dv *openShiftClusterDynamicValidator) validateProviders(ctx context.Context, providerClient resources.ProvidersClient) error {
+	providers := []string{
+		"Microsoft.Storage",
+		"Microsoft.Authorization",
+		"Microsoft.Compute",
+		"Microsoft.Network",
+	}
+
+	results, err := providerClient.List(ctx, nil, "")
+	if err != nil {
+		return err
+	}
+	for results.NotDone() {
+		for _, rp := range results.Values() {
+			for _, provider := range providers {
+				if provider == *rp.Namespace && *rp.RegistrationState != "Registered" {
+					return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidResourceProviderPermissions, "", "The resource provider '%s' is not registered.", provider)
+				}
+			}
+		}
+		results.Next()
+	}
```

Comment:
the page iteration from List needs to go into pkg/util/azureclient/mgmt/resources/providers_addons.go.  This will simplify unit testing.  Then iterate the slice returned from List here.